### PR TITLE
Copy lora functions to XLPipelines

### LINKED ISSUE #4511: StableDiffusionXLInpaintPipeline and StableDiffusionXLImg2ImgPipeline
I have tested using the `load_lora_weights()` and `save_lora_weights()` functions from `StableDiffusionXLPipeline` in `StableDiffusionXLInpaintPipeline` and `StableDiffusionXLImg2ImgPipeline`, and they work well. 

Can I open a PR?

---

              This is likely happening because the definitions of `load_lora_weights()` and `save_lora_weights()` (irrelevant here but for comprehensiveness) are not there in the pipeline source. 

https://github.com/huggingface/diffusers/blob/816ca0048f7a38531a6f036299d2dd274abb5639/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L866

https://github.com/huggingface/diffusers/blob/816ca0048f7a38531a6f036299d2dd274abb5639/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L898

I think adding the `load_lora_weights()` method to `StableDiffusionXLImg2ImgPipeline` should solve the problem. 

If possible for you, would you maybe like to fix this and open a PR?

_Originally posted by @sayakpaul in https://github.com/huggingface/diffusers/issues/4438#issuecomment-1662336456_

### PR DESCRIPTION
**Pull request by wooyeolBaek**
# What does this PR do?

Fixes #4511 

Add `load_lora_weights()` and `save_lora_weights()` from `StableDiffusionXLPipeline` to `StableDiffusionXLInpaintPipeline` and `StableDiffusionXLImg2ImgPipeline`

```python
# Overrride to properly handle the loading and unloading of the additional text encoder.
def load_lora_weights(self, pretrained_model_name_or_path_or_dict: Union[str, Dict[str, torch.Tensor]], **kwargs):
    # We could have accessed the unet config from `lora_state_dict()` too. We pass
    # it here explicitly to be able to tell that it's coming from an SDXL
    # pipeline.
    state_dict, network_alphas = self.lora_state_dict(
        pretrained_model_name_or_path_or_dict,
        unet_config=self.unet.config,
        **kwargs,
    )
    self.load_lora_into_unet(state_dict, network_alphas=network_alphas, unet=self.unet)

    text_encoder_state_dict = {k: v for k, v in state_dict.items() if "text_encoder." in k}
    if len(text_encoder_state_dict) > 0:
        self.load_lora_into_text_encoder(
            text_encoder_state_dict,
            network_alphas=network_alphas,
            text_encoder=self.text_encoder,
            prefix="text_encoder",
            lora_scale=self.lora_scale,
        )

    text_encoder_2_state_dict = {k: v for k, v in state_dict.items() if "text_encoder_2." in k}
    if len(text_encoder_2_state_dict) > 0:
        self.load_lora_into_text_encoder(
            text_encoder_2_state_dict,
            network_alphas=network_alphas,
            text_encoder=self.text_encoder_2,
            prefix="text_encoder_2",
            lora_scale=self.lora_scale,
        )

@classmethod
def save_lora_weights(
    self,
    save_directory: Union[str, os.PathLike],
    unet_lora_layers: Dict[str, Union[torch.nn.Module, torch.Tensor]] = None,
    text_encoder_lora_layers: Dict[str, Union[torch.nn.Module, torch.Tensor]] = None,
    text_encoder_2_lora_layers: Dict[str, Union[torch.nn.Module, torch.Tensor]] = None,
    is_main_process: bool = True,
    weight_name: str = None,
    save_function: Callable = None,
    safe_serialization: bool = False,
):
    state_dict = {}

    def pack_weights(layers, prefix):
        layers_weights = layers.state_dict() if isinstance(layers, torch.nn.Module) else layers
        layers_state_dict = {f"{prefix}.{module_name}": param for module_name, param in layers_weights.items()}
        return layers_state_dict

    state_dict.update(pack_weights(unet_lora_layers, "unet"))

    if text_encoder_lora_layers and text_encoder_2_lora_layers:
        state_dict.update(pack_weights(text_encoder_lora_layers, "text_encoder"))
        state_dict.update(pack_weights(text_encoder_2_lora_layers, "text_encoder_2"))

    self.write_lora_layers(
        state_dict=state_dict,
        save_directory=save_directory,
        is_main_process=is_main_process,
        weight_name=weight_name,
        save_function=save_function,
        safe_serialization=safe_serialization,
    )

def _remove_text_encoder_monkey_patch(self):
    self._remove_text_encoder_monkey_patch_classmethod(self.text_encoder)
    self._remove_text_encoder_monkey_patch_classmethod(self.text_encoder_2)

```

Now you can use like this
```python
pipe = StableDiffusionXLImg2ImgPipeline.from_pretrained(args.base_model_id, torch_dtype=torch.float16)
pipe = pipe.to("cuda:1")
pipe.load_lora_weights(lora_weights_path)
generator = torch.Generator("cuda:1").manual_seed(0)
image = pipe(image=init_image, prompt=prompt, generator=generator).images[0]
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
  - https://github.com/huggingface/diffusers/issues/4511
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@sayakpaul

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_img2img.py [hunk 1]
```diff
@@ -1012,3 +1013,73 @@ def denoising_value_valid(dnv):
             return (image,)
 
         return StableDiffusionXLPipelineOutput(images=image)
+
+    # Overrride to properly handle the loading and unloading of the additional text encoder.
+    def load_lora_weights(self, pretrained_model_name_or_path_or_dict: Union[str, Dict[str, torch.Tensor]], **kwargs):
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_img2img.py:1018]: We can use `# Copied from ...` statements here and elsewhere where we're verbatim copying methods. 

Example:

https://github.com/huggingface/diffusers/blob/aef11cbf667ce5175fdca2937a8ab7df385ebe49/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L67

Comment (wooyeolBaek) [reply src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_img2img.py:1019]: `# Copied from diffusers.pipelines.stable_diffusion_xl.pipeline_stable_diffusion_xl.load_lora_weights`
Thanks, I just left `Copy from ...` statement in front of the functions.

**Comment by sayakpaul** [comment]
@wooyeolBaek there still seems to be a mismatch as per the CI logs. Could you run `make fix-copies` and push the changes?

**Comment by wooyeolBaek** [comment]
> @wooyeolBaek there still seems to be a mismatch as per the CI logs. Could you run `make fix-copies` and push the changes?

Thanks, I think it's resolved now.

**Review by sayakpaul — APPROVED** [review]
Gracias!

**Pull request merged**